### DISCUSSION TITLE
Expression tree binds to interface method when using generics

### DISCUSSION THREAD

**Discussion by koenbeuk** [dotnet/roslyn#58566] [Q&A] [answered]
Given:
```csharp
var expression = FromExpression<Foo>();

Expression<Func<TFoo, int>> FromExpression<TFoo>() where TFoo : IFoo
  => foo => foo.GetBar();

interface IFoo {
     int GetBar();
}

class Foo : IFoo { 
    int GetBar() => 0;
}
```

Given that we're using generics, I would assume that the expression `foo.GetBar()` would bind to `Foo::GetBar()` instead it turns out this binds to `IFoo.GetBar()`

Is this a bug or by design? it feels counter intuitive. I'm using VS 2022 preview with dotnet 6

Looking forward to any input!

**Comment by CyrusNajmabadi** [comment, marked as answer]
This is by design.  The expression tree has no concept of the calling point.  It is compiled and relative to the creation point.  e.g.:

```c#
... where TFoo : IFoo
  => foo => foo.GetBar();
```

So this `foo` is only known to be an IFoo.

Note: i would not expect your code to compile. `Foo` doesn't implement `IFoo`, so it's not legal to use in `FromExpression<Foo>` as it won't satisfy the constraint.

**Reply by koenbeuk** [reply]
Thanks! (And fixed the sample such that Foo now implements IFoo correctly)

**Reply by CyrusNajmabadi** [reply]
Sure :)

DId you understand the answer above?  It's worthwhile to recognize that 'generics' are not 'templates'.  We don't replace the body textually with the types you specified.  INstead, the method is bound independent of the instantiation and always has that singular meaning.  Then, the instantiation is just used for the behind-the-scenes codegen, and optimizations are performed there.  HOwever, from a static-semantics level, it changes nothing.  So, for this, regardless of hwat you instantiate as, the static semantics are that it is calling through the interface method.

**Reply by koenbeuk** [reply]
I'm struggling a bit with the understand part:

> the method is bound independent of the instantiation and always has that singular meaning

This is not the case with non-expression generic methods correct? If I had: `int GetBar<TFoo>(TFoo foo) : where TFoo : IFoo => tfoo.GetBar()` then it would compile to JIT that binds GetBar to whatever TFoo is.

**Reply by CyrusNajmabadi** [reply]
> This is not the case with non-expression generic methods correct? 

Incorrect :)

It is the case with normal generic methods as well.

> `int GetBar<TFoo>(TFoo foo) : where TFoo : IFoo => tfoo.GetBar()` then it would compile to JIT 

I think i see the confusion.  There are *two* compile steps here.  The first is the language.  This encodes static semantics into IL in a well defined fashion.  From the language's (and the IL's) perspective, this is an interface dispatch through that particular type-parameter type.

Now, there is a second compile step that happens later on (normally by the JIT).  Here, the JIT takes that IL and decides what to do with it.  First, this is completely in the air as to how this happens (it's effectively an implementation detail).  In practice, the standard MS runtime does the following:

1. for all *value type* instantiations, you get a distinct copy of code generated that uses just that value type.
2. all *reference* type instantiates *share* a single copy of the code.

However, that policy doesn't even really matter as far as your question is concerned.  Even if each different reference type got a different instantiation, it's still going to call through the interface.  Why?  Because that's literally what the IL says to do.  :)

**Comment by koenbeuk** [comment]
In case others deal with this scenario, I worked around it by passing the expression through a visitor that rewrites the method calls, e.g.:

```csharp
    sealed class GetBarMethodReplacer<TFoo> : ExpressionVisitor
    {
        static MethodInfo _replaceMethodInfo;
        static MethodInfo _replaceWithMethodInfo;

        static GetSpaceIdMethodReplacer()
        {
            _replaceMethodInfo = typeof(IFoo).GetMethods()
                .Where(x => x.Name == nameof(IFoo.GetBar))
                .Single();

            _replaceWithMethodInfo = typeof(TFoo).GetMethods()
                .Where(x => x.Name == nameof(TFoo.GetBar))
                .Single();
        }

        protected override Expression VisitMethodCall(MethodCallExpression node)
        {
            if (node.Method == _replaceMethodInfo)
            {
                return Expression.Call(
                    Visit(node.Object), 
                    _replaceWithMethodInfo, 
                    Visit(node.Arguments));
            }

            return base.VisitMethodCall(node);
        }
    }
```
